graph/path/yen_ksp.go

```diff
@@ -0,0 +1,156 @@
+// Copyright ©2018 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package path
+
+import (
+	"sort"
+
+	"gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/graph/internal/set"
+)
+
+func YenKShortestPath(g graph.Graph, k int, s, t graph.Node) [][]graph.Node {
+	yk := yenKSPAdjuster{
+		g:       g,
+		visited: make(map[int64]set.Int64s),
+	}
+
+	var weight Weighting
+	if wg, ok := g.(Weighted); ok {
+		weight = wg.Weight
+	} else {
+		weight = UniformCost(g)
+	}
+
+	yk.weight = weight
+
+	var paths [][]graph.Node
+
+	shortest, _ := DijkstraFrom(s, yk).To(t.ID())
+	if len(shortest) != 0 {
+		paths = append(paths, shortest)
+	} else {
+		return paths
+	}
+
+	var pot []yenShortest
+
+	for i := int64(1); i < int64(k); i++ {
+		for n := 0; n < (len(paths[i-1]) - 1); n++ {
+			spur := paths[i-1][n]
+			root := make([]graph.Node, len(paths[i-1][:n+1]))
+			copy(root, paths[i-1][:n+1])
+
+			var rootWeight float64
+			for x := 1; x < len(root); x++ {
+				w, _ := weight(root[x-1].ID(), root[x].ID())
+				rootWeight += w
+			}
+
+			for _, path := range paths {
+				if len(path) > n {
+					ok := true
+					for x := 0; x < len(root); x++ {
+						if path[x].ID() != root[x].ID() {
+							ok = false
+							break
+						}
+					}
+					if ok {
+						yk.addVisited(path[n].ID(), path[n+1].ID())
+					}
+				}
+			}
+
+			spath, weight := DijkstraFrom(spur, yk).To(t.ID())
+			size := len(root) - 1
+
+			if len(root) > 1 {
+				nroot := root[:size]
+				nroot = append(nroot, spath...)
+				potential := yenShortest{nroot, weight + rootWeight}
+				pot = append(pot, potential)
+			} else {
+				potential := yenShortest{spath, weight}
+				pot = append(pot, potential)
+			}
+
+			yk.visited = make(map[int64]set.Int64s)
+		}
+
+		if len(pot) == 0 {
+			break
+		}
+
+		sort.Sort(byPathWeight(pot))
+
+		paths = append(paths, pot[0].p)
+
+		pot = pot[1:]
+	}
+
+	return paths
+}
+
+type yenShortest struct {
+	p      []graph.Node
```

Comment:
s/p/path/g

Reply:
What do you mean by this?

Reply:
Please replace the `p` field label with `path`. The name exists in enough locations at a distance from its definition, that a richer name is worthwhile.